### PR TITLE
Added sy.Module

### PR DESCRIPTION
**Pull request by madhavajay**
## Description
Added sy.Module which aims to provide similar functionality to nn.Module

## Affected Dependencies
None

## How has this been tested?
Locally and CI

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Review by tudorcebere — APPROVED** [review]
Awesome!

File: src/syft/core/common/module.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# stdlib
+from collections import OrderedDict
+from typing import Any
+from typing import List
+from typing import Optional
+from typing import Union
+
+
+# circular imports when using the syft.lib.full_name_with_qualname version
```
Comment (Jasopaum) [inline src/syft/core/common/module.py:9]: Really? Isn't `full_name_with_qualname` in `src/syft/lib` which doesn't much?

File: src/syft/core/common/module.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+def full_name_with_qualname(klass: type) -> str:
+    return f"{klass.__module__}.{klass.__qualname__}"
+
+
+class Module:
```
Comment (Jasopaum) [inline src/syft/core/common/module.py:14]: Should we put this in the torch folder if it's torch specific?

File: tests/syft/core/common/module_test.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            x = torch.flatten(x, 1)
+            x = self.fc1(x)
+            x = F.relu(x)
+            x = self.dropout2(x)
+            x = self.fc2(x)
+            output = F.log_softmax(x, dim=1)
+            return output
+
+    vanilla_model = Net()
+    sy_model = SyNet()  # without the k it kant take over the world
```
Comment (gmuraru) [inline tests/syft/core/common/module_test.py:110]: TYPO

File: src/syft/lib/torch/allowlist.py [hunk 4]
```diff
@@ -667,89 +667,90 @@
...
 
 # Optimizers
-# allowlist["torch.optim.Adadelta"] = "torch.optim.Adadelta"
-# allowlist["torch.optim.lr_scheduler.StepLR"] = "torch.optim.lr_scheduler.StepLR"
-# allowlist["torch.optim.lr_scheduler.StepLR.step"] = "syft.lib.python.SyNone"
-# allowlist["torch.optim.Adadelta.zero_grad"] = "syft.lib.python.SyNone"
-# allowlist["torch.optim.Adadelta.step"] = "syft.lib.python.SyNone"
+allowlist["torch.optim.Adadelta"] = "torch.optim.Adadelta"
+allowlist["torch.optim.lr_scheduler.StepLR"] = "torch.optim.lr_scheduler.StepLR"
+allowlist["torch.optim.lr_scheduler.StepLR.step"] = "syft.lib.python._SyNone"
```
Comment (gmuraru) [inline src/syft/lib/torch/allowlist.py:748]: Q: What is ```_SyNone```?

**Review by gmuraru — APPROVED** [review]

File: tests/syft/core/common/module_test.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        assert model.training is True
+
+        model.eval()  # type: ignore
+
+        assert model.training is False
+
+    # assert vanilla_model.modules == sy_model._modules
+    assert len(vanilla_model.modules) == len(sy_model.modules)
+    assert len(sy_model.modules) == 6
+    for _, module in sy_model.modules.items():
```
Comment (Jasopaum) [inline tests/syft/core/common/module_test.py:136]: nit: I think
```
for syft_module, vanilla_module in zip(sy_model.modules.values(), vanilla_model.modules.values())
```
or something like that would be better

File: src/syft/lib/torch/allowlist.py (hunk 4 above)
Comment (madhavajay) [reply src/syft/lib/torch/allowlist.py:748]: That is the renamed Constructor Class for SyNone which is the shared instance variable.

File: tests/syft/core/common/module_test.py (hunk 3 above)
Comment (madhavajay) [reply tests/syft/core/common/module_test.py:110]: Its a dad joke 😂 The K is for SkyNet... 😬 But I can see this tripping up a spell checker.

File: src/syft/core/common/module.py (hunk 1 above)
Comment (madhavajay) [reply src/syft/core/common/module.py:9]: Yeah I tried, so weird. I guess its the other way around and the sy.Module itself is causing the issue?
Screen Shot 2020-10-29 at 8 40 54 am

File: src/syft/core/common/module.py (hunk 2 above)
Comment (madhavajay) [reply src/syft/core/common/module.py:14]: Yes open to suggestions. I have created a seperate ticket: https://github.com/OpenMined/PySyft/issues/4747

File: tests/syft/core/common/module_test.py (hunk 5 above)
Comment (madhavajay) [reply tests/syft/core/common/module_test.py:136]: Fixed.

**Comment by madhavajay** [comment]
This got merged here:
https://github.com/OpenMined/PySyft/pull/4696
